### PR TITLE
Add 3D attention mask documentation to Transformer

### PR DESCRIPTION
**Pull request by UsmannK**
In #31996 (committed in db02a4e4ceced885508991441ea68896d48fc369) we added support for 3-dimensional attention masks to the MultiheadAttention layer. This PR updates the Transformer documentation to reflect this. The Transformer{Encoder,Decoder,EncoderLayer,DecoderLayer} class documentation does not need any changes as they never referenced the shapes to begin with. I also fixed some wording and a LaTeX typo in the MultiheadAttention docs.

Here is some sample code demonstrating the newly documented functionality:

```
import torch
N = 3
S = 20
T = 1
E = 4
nhead = 2
transformer = torch.nn.Transformer(d_model=E, nhead=nhead)

src = torch.randn(S, N, E)
tgt = torch.randn(T, N, E)
src_mask = torch.randn(N*nhead, S, S)
tgt_mask = torch.randn(N*nhead, T, T)
memory_mask = torch.randn(N*nhead, T, S)

print(transformer(src, tgt, src_mask=src_mask, tgt_mask=tgt_mask, memory_mask=memory_mask))
```
```
tensor([[[ 1.2656,  0.0161,  0.2470, -1.5287],
         [-1.6573,  0.9786,  0.5202,  0.1585],
         [-0.8682,  1.5458,  0.2211, -0.8987]]],
       grad_fn=<NativeLayerNormBackward>)
```

### CODE REVIEW

File: torch/nn/modules/activation.py [hunk 1]
```diff
@@ -891,13 +891,12 @@ def forward(self, query, key, value, key_padding_mask=None,
...
           with the zero positions will be unchanged. If a BoolTensor is provided, the positions with the
           value of ``True`` will be ignored while the position with the value of ``False`` will be unchanged.
-        - attn_mask: 2D mask :math:`(L, S)` where L is the target sequence length, S is the source sequence length.
-          3D mask :math:`(N*num_heads, L, S)` where N is the batch size, L is the target sequence length,
-          S is the source sequence length. attn_mask ensure that position i is allowed to attend the unmasked
-          positions. If a ByteTensor is provided, the non-zero positions are not allowed to attend
-          while the zero positions will be unchanged. If a BoolTensor is provided, positions with ``True``
-          is not allowed to attend while ``False`` values will be unchanged. If a FloatTensor
-          is provided, it will be added to the attention weight.
+        - attn_mask: Either a 2D mask :math:`(L, S)` or a 3D mask :math:`(N*num\_heads, L, S)` where N is the
```
Comment (zhangguanheng66) [inline torch/nn/modules/activation.py:894]: `num\_heads`?

File: torch/nn/modules/transformer.py [hunk 2]
```diff
@@ -84,9 +84,9 @@ def forward(self, src: Tensor, tgt: Tensor, src_mask: Optional[Tensor] = None, t
         Shape:
             - src: :math:`(S, N, E)`.
             - tgt: :math:`(T, N, E)`.
-            - src_mask: :math:`(S, S)`.
-            - tgt_mask: :math:`(T, T)`.
-            - memory_mask: :math:`(T, S)`.
+            - src_mask: :math:`(S, S)` or :math:`(N*nhead, S, S)`.
```
Comment (zhangguanheng66) [inline torch/nn/modules/transformer.py:87]: Better to be consistent between `nhead` and `num_heads`.

**Review by zhangguanheng66 — COMMENTED** [review]
IMO, we just need an update for the shape doc.

File: torch/nn/modules/activation.py (hunk 1 above)
Comment (UsmannK) [reply torch/nn/modules/activation.py:894]: We need to escape the underscore otherwise it looks like this: https://pytorch.org/docs/master/generated/torch.nn.MultiheadAttention.html

image

File: torch/nn/modules/transformer.py (hunk 2 above)
Comment (UsmannK) [reply torch/nn/modules/transformer.py:87]: Definitely. I went with what was already there. The Transformer documentation uses `nhead` https://pytorch.org/docs/master/generated/torch.nn.Transformer.html

The MultiHeadAttention docs use `num_heads`. https://pytorch.org/docs/master/generated/torch.nn.MultiheadAttention.html.

I did not diverge from these.

**Comment by UsmannK** [comment]
> IMO, we just need an update for the shape doc.

Sorry, not sure what you mean by this.

File: torch/nn/modules/transformer.py (hunk 2 above)
Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py:87]: For MHA, we should use `num_heads`.

File: torch/nn/modules/activation.py (hunk 1 above)
Comment (zhangguanheng66) [inline torch/nn/modules/activation.py:894]: I mean this part of change is not necessary, IMO.

File: torch/nn/modules/transformer.py (hunk 2 above)
Comment (UsmannK) [reply torch/nn/modules/transformer.py:87]: Yes I use `num_heads` in MHA. You'll notice that I use `nhead` in transformer.py and `num_heads` in activation.py

File: torch/nn/modules/activation.py (hunk 1 above)
Comment (UsmannK) [reply torch/nn/modules/activation.py:894]: I think this section of the documentation was unclear and a little hard to read before. Also it has a LaTeX typo as shown here: https://github.com/pytorch/pytorch/pull/41205#discussion_r453871147
